File: frontend/scripts/openings-golden.test.ts
// DEC-9 golden 对拍（TS 侧）：共享 fixtures 位于 api/golden/openings/，
// Go 权威实现（backend/internal/schedule）与前端 TS 实现必须对同一批 fixtures
// 给出逐字段一致的结果。本测试锁住「fixtures 是当前 TS 实现的诚实快照」——
// 防止 fixtures 被手工篡改或 TS 实现漂移后忘记再生成。
import assert from 'node:assert/strict'
import { readdirSync, readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'
import test from 'node:test'

import { buildCalendarModel, calculateMonthOverview } from '../src/pages/calendar/model.ts'
import type { ScheduleAvailability, ScheduleSlotListItem } from '../src/api/client.ts'

const GOLDEN_DIR = join(dirname(fileURLToPath(import.meta.url)), '..', '..', 'api', 'golden', 'openings')

interface GoldenSlot {
  id: string
  type: 'shoot' | 'hold' | 'busy'
  start_at: string
  end_at: string
  order_status: string | null
}

interface GoldenFixture {
  name: string
  description: string
  timezone: string
  availability: ScheduleAvailability
  slots: GoldenSlot[]
  dates: string[]
  month: string | null
  expected: {
    days: Array<{
      date: string
      working_window: { start: string; end: string; start_at: string; end_at: string } | null
      openings: Array<{ date: string; start: string; end: string; start_at: string; end_at: string }>
    }>
    errors: Array<{ date: string; message: string }>
    month_overview: {
      shoot_count: number
      hold_days: number
      conflict_days: number
      open_days: number
      utilization: number | null
    } | null
  }
}

function loadFixtures(): GoldenFixture[] {
  const files = readdirSync(GOLDEN_DIR)
    .filter((name) => name.endsWith('.json'))
    .sort()
  // 必备用例名单：DST/月利用率等关键语义被误删且未重生成时，守护测试直接失败。
  for (const required of [
    'basic-weekday.json',
    'dst-window-resolution.json',
    'openings-boundary.json',
    'month-utilization.json',
  ]) {
    if (!files.includes(required)) {
      assert.fail(`golden fixture ${required} 缺失：关键语义用例不得删除，需求变更走 gen-openings-golden.ts 重生成`)
    }
  }
  return files.map((name) => JSON.parse(readFileSync(join(GOLDEN_DIR, name), 'utf8')) as GoldenFixture)
}

test('openings golden fixtures 与当前 TS 实现逐字段一致（DEC-9）', () => {
  const fixtures = loadFixtures()
  assert.ok(fixtures.length >= 4, 'golden fixtures 至少应覆盖 4 个用例文件')
  for (const fixture of fixtures) {
    const slots = fixture.slots.map((slot) => slot as unknown as ScheduleSlotListItem)
    const model = buildCalendarModel(slots, fixture.dates, fixture.timezone, fixture.availability)
    const actualDays = model.days.map((day) => ({
      date: day.date,
      working_window: day.workingWindow
        ? {
            start: day.workingWindow.start,
            end: day.workingWindow.end,
            start_at: day.workingWindow.startAt,
            end_at: day.workingWindow.endAt,
          }
        : null,
      openings: day.openings.map((opening) => ({
        date: opening.date,
        start: opening.start,
        end: opening.end,
        start_at: opening.startAt,
        end_at: opening.endAt,
      })),
    }))
    assert.deepEqual(actualDays, fixture.expected.days, `${fixture.name}: days 漂移`)

    let monthOverview: GoldenFixture['expected']['month_overview'] = null
    if (fixture.month) {
      const { overview } = calculateMonthOverview(slots, fixture.month, fixture.timezone, fixture.availability)
      monthOverview = {
        shoot_count: overview.shootCount,
        hold_days: overview.holdDays,
        conflict_days: overview.conflictDays,
        open_days: overview.openDays,
        utilization: overview.utilization,
      }
    }
    assert.deepEqual(monthOverview, fixture.expected.month_overview, `${fixture.name}: month_overview 漂移`)
    assert.deepEqual(model.errors, fixture.expected.errors, `${fixture.name}: errors 漂移`)
  }
})

File: backend/internal/schedule/openings_golden_test.go
package schedule

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strconv"
	"testing"
	"time"
)

// DEC-9 golden 对拍（Go 侧）：共享 fixtures 位于 api/golden/openings/，
// 由前端 TS 实现（frontend/src/pages/calendar/model.ts + scripts/gen-openings-golden.ts）
// 生成。Go 权威实现必须与 TS 对同一批 fixtures 给出逐字段一致的结果；
// fixtures 变更须经生成脚本再生成，禁止手改期望值迁就实现。

const goldenFixturesDir = "../../../api/golden/openings"

type goldenWindowSpec struct {
	Start string `json:"start"`
	End   string `json:"end"`
}

type goldenSlot struct {
	ID          string  `json:"id"`
	Type        string  `json:"type"`
	StartAt     string  `json:"start_at"`
	EndAt       string  `json:"end_at"`
	OrderStatus *string `json:"order_status"`
}

type goldenDayExpected struct {
	Date          string            `json:"date"`
	WorkingWindow *goldenDayWindow  `json:"working_window"`
	Openings      []goldenOpeningEV `json:"openings"`
}

type goldenDayWindow struct {
	Start   string `json:"start"`
	End     string `json:"end"`
	StartAt string `json:"start_at"`
	EndAt   string `json:"end_at"`
}

type goldenOpeningEV struct {
	Date    string `json:"date"`
	Start   string `json:"start"`
	End     string `json:"end"`
	StartAt string `json:"start_at"`
	EndAt   string `json:"end_at"`
}

type goldenMonthOverview struct {
	ShootCount   int      `json:"shoot_count"`
	HoldDays     int      `json:"hold_days"`
	ConflictDays int      `json:"conflict_days"`
	OpenDays     int      `json:"open_days"`
	Utilization  *float64 `json:"utilization"`
}

type goldenFixture struct {
	Name         string `json:"name"`
	Timezone     string `json:"timezone"`
	Availability struct {
		Weekly            map[string]*goldenWindowSpec `json:"weekly"`
		MinOpeningMinutes int                          `json:"min_opening_minutes"`
	} `json:"availability"`
	Slots    []goldenSlot `json:"slots"`
	Dates    []string     `json:"dates"`
	Month    *string      `json:"month"`
	Expected struct {
		Days          []goldenDayExpected  `json:"days"`
		Errors        []goldenDayErrorEV   `json:"errors"`
		MonthOverview *goldenMonthOverview `json:"month_overview"`
	} `json:"expected"`
}

type goldenDayErrorEV struct {
	Date    string `json:"date"`
	Message string `json:"message"`
}

func loadGoldenFixtures(t *testing.T) []goldenFixture {
	t.Helper()
	entries, err := os.ReadDir(goldenFixturesDir)
	if err != nil {
		t.Fatalf("read golden fixtures dir: %v", err)
	}
	names := make([]string, 0, len(entries))
	for _, entry := range entries {
		if filepath.Ext(entry.Name()) == ".json" {
			names = append(names, entry.Name())
		}
	}
	// 必备用例名单：DST/月利用率等关键语义被误删且未重生成时，守护测试直接失败。
	for _, required := range []string{
		"basic-weekday.json", "dst-window-resolution.json", "openings-boundary.json", "month-utilization.json",
	} {
		found := false
		for _, name := range names {
			if name == required {
				found = true
				break
			}
		}
		if !found {
			t.Fatalf("golden fixture %s 缺失：关键语义用例不得删除，需求变更走 scripts/gen-openings-golden.ts 重生成", required)
		}
	}
	fixtures := make([]goldenFixture, 0, len(names))
	for _, name := range names {
		raw, err := os.ReadFile(filepath.Join(goldenFixturesDir, name))
		if err != nil {
			t.Fatalf("read fixture %s: %v", name, err)
		}
		var fixture goldenFixture
		if err := json.Unmarshal(raw, &fixture); err != nil {
			t.Fatalf("parse fixture %s: %v", name, err)
		}
		fixtures = append(fixtures, fixture)
	}
	return fixtures
}

func goldenPlan(f goldenFixture) AvailabilityPlan {
	plan := AvailabilityPlan{
		Weekly:            make(map[int]*DailyWindow, len(f.Availability.Weekly)),
		MinOpeningMinutes: f.Availability.MinOpeningMinutes,
	}
	for key, window := range f.Availability.Weekly {
		iso, err := strconv.Atoi(key)
		if err != nil {
			continue
		}
		if window == nil {
			plan.Weekly[iso] = nil
			continue
		}
		plan.Weekly[iso] = &DailyWindow{Start: window.Start, End: window.End}
	}
	return plan
}

func parseGoldenInstant(t *testing.T, s string) time.Time {
	t.Helper()
	parsed, err := time.Parse(time.RFC3339, s)
	if err != nil {
		t.Fatalf("parse instant %s: %v", s, err)
	}
	return parsed
}

func TestOpeningsGoldenParityWithTS(t *testing.T) {
	for _, fixture := range loadGoldenFixtures(t) {
		t.Run(fixture.Name, func(t *testing.T) {
			loc, err := time.LoadLocation(fixture.Timezone)
			if err != nil {
				t.Fatalf("load timezone: %v", err)
			}
			plan := goldenPlan(fixture)
			slots := make([]DaySlot, 0, len(fixture.Slots))
			for _, slot := range fixture.Slots {
				slots = append(slots, DaySlot{
					ID:          slot.ID,
					Type:        slot.Type,
					StartAt:     parseGoldenInstant(t, slot.StartAt),
					EndAt:       parseGoldenInstant(t, slot.EndAt),
					OrderStatus: slot.OrderStatus,
				})
			}

			days := make([]goldenDayExpected, 0, len(fixture.Dates))
			var dayErrors []goldenDayErrorEV
			for _, dateStr := range fixture.Dates {
				date, err := time.Parse("2006-01-02", dateStr)
				if err != nil {
					t.Fatalf("parse date %s: %v", dateStr, err)
				}
				window, err := ResolveDayWindow(date, loc, plan)
				if err != nil {
					dayErrors = append(dayErrors, goldenDayErrorEV{Date: dateStr, Message: err.Error()})
					days = append(days, goldenDayExpected{Date: dateStr, WorkingWindow: nil, Openings: []goldenOpeningEV{}})
					continue
				}
				expectedWindow := (*goldenDayWindow)(nil)
				if window != nil {
					expectedWindow = &goldenDayWindow{
						Start:   window.StartAt.In(loc).Format("15:04"),
						End:     window.EndAt.In(loc).Format("15:04"),
						StartAt: window.StartAt.UTC().Format(time.RFC3339),
						EndAt:   window.EndAt.UTC().Format(time.RFC3339),
					}
				}
				openings := DayOpenings(date, window, slots, plan.MinOpeningMinutes, loc)
				openingEVs := make([]goldenOpeningEV, 0, len(openings))
				for _, opening := range openings {
					openingEVs = append(openingEVs, goldenOpeningEV{
						Date:    opening.Date.Format("2006-01-02"),
						Start:   opening.StartAt.In(loc).Format("15:04"),
						End:     opening.EndAt.In(loc).Format("15:04"),
						StartAt: opening.StartAt.UTC().Format(time.RFC3339),
						EndAt:   opening.EndAt.UTC().Format(time.RFC3339),
					})
				}
				days = append(days, goldenDayExpected{Date: dateStr, WorkingWindow: expectedWindow, Openings: openingEVs})
			}
			if !reflect.DeepEqual(days, fixture.Expected.Days) {
				t.Fatalf("days 与 TS golden 不一致:\ngot  %+v\nwant %+v", days, fixture.Expected.Days)
			}
			errorsEV := fixture.Expected.Errors
			if len(dayErrors) == 0 {
				// 统一 nil/空比较：TS 侧序列化为 []。
				dayErrors = []goldenDayErrorEV{}
			}
			if !reflect.DeepEqual(dayErrors, errorsEV) {
				t.Fatalf("errors 与 TS golden 不一致: got %+v want %+v", dayErrors, errorsEV)
			}

			if fixture.Month != nil {
				monthStart, err := time.Parse("2006-01-02", *fixture.Month+"-01")
				if err != nil {
					t.Fatalf("parse month %s: %v", *fixture.Month, err)
				}
				days := DaysInMonth(monthStart)
				overview, overviewErrs := MonthOverview(slots, monthStart, days, loc, plan)
				_ = overviewErrs // 月概览错误已在逐日 errors 面对拍；这里不重复断言
				got := goldenMonthOverview(overview)
				if !reflect.DeepEqual(got, *fixture.Expected.MonthOverview) {
					t.Fatalf("month_overview 与 TS golden 不一致: got %+v want %+v", got, *fixture.Expected.MonthOverview)
				}
			}
		})
	}
}
